verifier/services/reorg_detector_service.go

```diff
@@ -0,0 +1,711 @@
+	}
+
+	blockAfterLCA := currentBlock
+
+	// Walk back until we find a matching block or hit finalized boundary
+	for parent.Number >= latestFinalizedBlock {
+		// Check if we have this block in our tail
+		r.tailMu.RLock()
+		ourBlock, exists := r.tailBlocks[parent.Number]
+		r.tailMu.RUnlock()
+
+		if exists && ourBlock.Hash == parent.Hash {
+			// Found LCA! Return blockAfterLCA
+			r.lggr.Debugw("Found matching block (LCA)",
+				"chainSelector", r.config.ChainSelector,
+				"lcaBlock", parent.Number,
+				"lcaHash", parent.Hash)
+			return blockAfterLCA, nil
+		}
+
+		// Continue walking back
+		blockAfterLCA = *parent
+		parent, err = r.sourceReader.GetBlockHeaderByHash(ctx, parent.ParentHash)
+		if err != nil {
+			return protocol.BlockHeader{}, fmt.Errorf("failed to fetch parent block by hash %s: %w", blockAfterLCA.ParentHash, err)
+		}
+		if parent == nil {
+			return protocol.BlockHeader{}, fmt.Errorf("parent block not found for hash %s", blockAfterLCA.ParentHash)
+		}
+	}
+
+	// Reorg is deeper than finalized block - finality violation
+	r.tailMu.RLock()
+	ourFinalizedBlock, exists := r.tailBlocks[latestFinalizedBlock]
+	r.tailMu.RUnlock()
+
+	return blockAfterLCA, &finalityViolationError{
+		violatedBlock:              ourFinalizedBlock,
+		newChainBlock:              blockAfterLCA,
+		finalizedNum:               latestFinalizedBlock,
+		latestFinalizedBlockExists: exists,
+	}
+}
+
+// rebuildTailFromBlock rebuilds the tail starting from the given block (block after LCA)
+// up to the current latest block.
+func (r *ReorgDetectorService) rebuildTailFromBlock(ctx context.Context, startBlock protocol.BlockHeader) error {
+	r.lggr.Infow("Rebuilding tail from block",
+		"chainSelector", r.config.ChainSelector,
+		"startBlock", startBlock.Number)
+
+	// Get current latest block to know how far to fetch
+	latest, _, err := r.headTracker.LatestAndFinalizedBlock(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get latest block: %w", err)
+	}
+	if latest == nil {
+		return fmt.Errorf("latest block is nil")
+	}
+
+	// Fetch blocks from startBlock to current latest
+	endBlock := latest.Number
+
+	var blockNumbers []*big.Int
+	for i := startBlock.Number; i <= endBlock; i++ {
+		blockNumbers = append(blockNumbers, new(big.Int).SetUint64(i))
+	}
+
+	headers, err := r.sourceReader.GetBlocksHeaders(ctx, blockNumbers)
+	if err != nil {
+		return fmt.Errorf("failed to fetch block headers: %w", err)
+	}
+
+	// Clear old tail and rebuild
+	r.tailMu.Lock()
+	defer r.tailMu.Unlock()
+
+	r.tailBlocks = make(map[uint64]protocol.BlockHeader)
+	for _, blockNum := range blockNumbers {
+		header, exists := headers[blockNum]
+		if !exists {
+			// Block might not exist yet if we're at chain tip
+			// In this case, just stop here
+			break
+		}
+		r.tailBlocks[header.Number] = header
+	}
+
+	r.tailMinBlock = startBlock.Number
+	r.tailMaxBlock = startBlock.Number + uint64(len(r.tailBlocks)) - 1
```

Comment:
Small nit but I think its clearer where the critical section is for this mutex if we update a local variables in these loops and explicitly lock/unlock after:


```suggestion
	// Clear old tail and rebuild
	tailBlocks = make(map[uint64]protocol.BlockHeader)
	for _, blockNum := range blockNumbers {
		header, exists := headers[blockNum]
		if !exists {
			// Block might not exist yet if we're at chain tip
			// In this case, just stop here
			break
		}
		tailBlocks[header.Number] = header
	}
	
	// Update local state
	r.tailMu.Lock()
	r.tailBlocks = tailBlocks
	r.tailMinBlock = startBlock.Number
	r.tailMaxBlock = startBlock.Number + uint64(len(r.tailBlocks)) - 1
	r.tailMu.Unlock()
```